&nbsp;A tabela de origem apresenta várias questões que precisam ser abordadas para conduzir uma análise significativa. Para esse fim, é imperativo que preparemos os dados, identificando e corrigindo quaisquer inconsistências, imprecisões ou informações ausentes. Depois que tivermos Abordamos esses problemas, selecionaremos e transformaremos cuidadosamente os elementos relevantes nos formatos apropriados para análise, garantindo que nossas idéias sejam precisas e acionáveis .            

In [1]:
import pandas as pd
import numpy as np
from string import ascii_letters

df = pd.read_json("../scrapy_project/output/data.jsonl", lines=True)
df.head(3)

model                                  Avionics:  \
0  Boeing 787-10 Dreamliner  Rockwell Collins next generation avionics   
1           Airbus A380-800            Thales Rockwell Honeywell Combo   
2             Airbus A300B2                                  Honeywell   

                  \n            Engine: \n            \
0  \n            2x Rolls-Royce Trent 1000 or Gen...   
1  \n            4x Engine Alliance GP7270 or Rol...   
2     \n            2x General Electric CF6, JT9D-7Q   

                         Power:          Max Cruise Speed:  \
0  76,000 pound-force            \n              515 knots   
1  81,500 pound-force            \n              587 knots   
2  53,000 pound-force            \n              495 knots   

                  Approach Speed (Vref):  \
0  \n              140 knots               
1  \n              135 knots               
2  \n              131 knots               

  \n              Travel range:\n              \
0          \n            6,430 Nautical Miles   
1          \n            8,000 Nautical Miles   
2          \n            1,850 Nautical Miles   

                               Fuel Economy:  \
0  \n            0.22 nautical mile / gallon   
1   \n            0.1 nautical mile / gallon   
2                   \n                         

                      Service Ceiling:                    Rate of Climb:  ...  \
0  \n            43,000 feet            \n            3000 feet / minute  ...   
1  \n            43,000 feet            \n            1500 feet / minute  ...   
2  \n            35,000 feet            \n            3000 feet / minute  ...   

              Seats - Business Class:                  Seats - First Class:  \
0  \n            290 seats                      \n            0               
1  \n            480 seats              \n            416 seats               
2           \n                                   \n                           

                                       Cabin Height:  \
0  \n              2.5 metre - 8.20 feet               
1    \n              3 metre - 9.84 feet               
2  \n              2.48 metre - 8.14 feet        ...   

                                      Cabin Width:  \
0  \n            5.49 metre - 18.01 feet             
1  \n            6.58 metre - 21.59 feet             
2  \n            5.13 metre - 16.83 feet             

                                       Cabin Length:  \
0     \n            62 metre - 203.41 feet             
1  \n            50.68 metre - 166.27 feet             
2   \n            40.7 metre - 133.53 feet             

                                    Exterior Length:  \
0   \n            68.3 metre - 224.08 feet             
1  \n            72.73 metre - 238.61 feet             
2  \n            53.61 metre - 175.88 feet             

                                       Tail height:  \
0  \n            17.02 metre - 55.84 feet             
1  \n            24.27 metre - 79.63 feet             
2  \n            16.67 metre - 54.69 feet             

                                Fuselage Diameter:  \
0  \n            5.77 metre - 18.93 feet             
1  \n            7.14 metre - 23.42 feet             
2   \n            5.4 metre - 17.72 feet             

                         Wing Span / Rotor Diameter:  \
0  \n            60.12 metre - 197.24 feet             
1  \n            79.75 metre - 261.64 feet             
2  \n            44.83 metre - 147.08 feet             

                               Wing Tips:  
0  \n            Raked Wingtips            
1  \n            Wingtip Fences            
2     \n            No Winglets            

[3 rows x 28 columns]

In [2]:
df[df.model=="McDonnell Douglas DC-8-72"]

model                                      Avionics:  \
10  McDonnell Douglas DC-8-72   Dual Collins VHF 21D,  HF 628T-1 ,  TDR-94D    

   \n            Engine: \n                                  Power:  \
10      \n            4x CFM 56-2 C6   34,000 pound-force             

            Max Cruise Speed:                 Approach Speed (Vref):  \
10  \n              488 knots  \n              145 knots               

   \n              Travel range:\n                         Fuel Economy:  \
10          \n            9,800 Nautical Miles  \n                         

                       Service Ceiling:                    Rate of Climb:  \
10  \n            42,000 feet            \n            2200 feet / minute   

    ...    Seats - Business Class:         Seats - First Class:  \
10  ...  \n                          \n                           

                                        Cabin Height:  \
10  \n              2.21 metre - 7.25 feet        ...   

                                       Cabin Width:  \
10  \n            3.51 metre - 11.52 feet             

                                        Cabin Length:  \
10  \n            42.49 metre - 139.40 feet             

                                  Exterior Length:  \
10  \n            48 metre - 157.48 feet             

                                       Tail height:  \
10  \n            13.1 metre - 42.98 feet             

                                 Fuselage Diameter:  \
10  \n            3.74 metre - 12.27 feet             

                         Wing Span / Rotor Diameter:  \
10  \n            42.3 metre - 138.78 feet             

                             Wing Tips:  
10  \n            No Winglets            

[1 rows x 28 columns]

In [3]:
#Tratamento para os nomes das colunas
old_cols = ["".join(l for l in i if l in set(ascii_letters+" ")).split() 
            for i in df.columns] 
new_cols = ["".join(w.capitalize() for w in i) for i in old_cols]
str_df = df.rename(columns={k:v for k,v in zip(df.columns,new_cols)})

#Remoção do excesso de espaços e identificação dos valores nulos
str_df = str_df.astype(str).applymap(lambda x: " ".join(w for w in x.split())) 
str_df = str_df.applymap(lambda x: "None" if x == "" else x)
str_df = str_df.applymap(lambda x: np.nan if x == "None" else x)

In [4]:
str_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 910 entries, 0 to 909
Data columns (total 28 columns):
 #   Column                 Non-Null Count  Dtype 
---  ------                 --------------  ----- 
 0   Model                  910 non-null    object
 1   Avionics               489 non-null    object
 2   Engine                 907 non-null    object
 3   Power                  910 non-null    object
 4   MaxCruiseSpeed         908 non-null    object
 5   ApproachSpeedVref      484 non-null    object
 6   TravelRange            884 non-null    object
 7   FuelEconomy            451 non-null    object
 8   ServiceCeiling         876 non-null    object
 9   RateOfClimb            814 non-null    object
 10  TakeOffDistance        610 non-null    object
 11  LandingDistance        591 non-null    object
 12  MaxTakeOffWeight       897 non-null    object
 13  MaxLandingWeight       435 non-null    object
 14  MaxPayload             874 non-null    object
 15  FuelTankCapacity       

In [5]:
#Seleção das colunas avaliadas em relação ao número de valores nulos
str_df = str_df.iloc[:,[0,2,3,4,6,8,9,12,14,15,23,26]].dropna().reset_index(drop=True)
str_df.head(3)

Model  \
0  Boeing 787-10 Dreamliner   
1           Airbus A380-800   
2             Airbus A300B2   

                                              Engine               Power  \
0  2x Rolls-Royce Trent 1000 or General Electric ...  76,000 pound-force   
1  4x Engine Alliance GP7270 or Rolls-Royce Trent...  81,500 pound-force   
2                   2x General Electric CF6, JT9D-7Q  53,000 pound-force   

  MaxCruiseSpeed           TravelRange ServiceCeiling         RateOfClimb  \
0      515 knots  6,430 Nautical Miles    43,000 feet  3000 feet / minute   
1      587 knots  8,000 Nautical Miles    43,000 feet  1500 feet / minute   
2      495 knots  1,850 Nautical Miles    35,000 feet  3000 feet / minute   

  MaxTakeOffWeight MaxPayload FuelTankCapacity             ExteriorLength  \
0       254,011 Kg  57,277 Kg    33,399 gallon   68.3 metre - 224.08 feet   
1       575,000 Kg  87,000 Kg    85,472 gallon  72.73 metre - 238.61 feet   
2       142,000 Kg  34,590 Kg    11,623 gallon  53.61 metre - 175.88 feet   

       WingSpanRotorDiameter  
0  60.12 metre - 197.24 feet  
1  79.75 metre - 261.64 feet  
2  44.83 metre - 147.08 feet

&nbsp;As colunas "ExteriorLength" e "WingSpanRotorDiameter" contêm dois conjuntos de medições, um em metros e outro em pés. Devido a essa característica única, precisaremos tratá-los de maneira diferente das outras colunas da tabela. Especificamente, precisaremos separar os dois conjuntos de medições e convertê-los em uma unidade de medição padronizada para consistência e facilidade de análise.

In [6]:
#Separação e tratamento da metade esquerda
le_df = str_df.iloc[:,2:-2]
le_df = le_df.applymap(lambda x: x.split()[0].replace(",","")).astype(float)

#Separação e tratamento da metade direita
ri_df = str_df.iloc[:,-2:]
ri_df = ri_df.applymap(lambda x: x.split()[-2]).astype(float)

#Junção das duas metades numa nova tabela
new_df = le_df.join(ri_df)

#Inserção dos modelos na nova tabela
new_df.insert(0,"Model",str_df.Model)

#Inserção do tipo de força do motor
power_unit = str_df.Power.map(lambda x: "lbf" if "pound-force" in x else "hp")
new_df.insert(2,"Unit",power_unit)

Algumas linhas da coluna "Engines" não contém a informação sobre quantos motores a aeronave possui.

In [7]:
#Localização dos modelos onde o número de motores não é informado 
engines = str_df.Engine.map(lambda x: x.split()[0][0])
str_df.loc[engines[engines.str.isnumeric()==False].index,"Model"]

67                 Cessna 404 Titan
294    Bombardier Q100 (Dash 8-100)
615          Mikoyan-Gurevich MiG-9
737              Boeing BBJ 787-800
Name: Model, dtype: object

Essas quatro aeronaves possuem 2 motores cada.

In [8]:
#Inserção da quantidade de motores na variável
engines[engines.str.isnumeric()==False] = "2"

#Inserção da variável com o número de motores na nova tabela 
new_df.insert(3,"EnginesQty",engines.astype(int))

In [9]:
new_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 759 entries, 0 to 758
Data columns (total 13 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   Model                  759 non-null    object 
 1   Power                  759 non-null    float64
 2   Unit                   759 non-null    object 
 3   EnginesQty             759 non-null    int64  
 4   MaxCruiseSpeed         759 non-null    float64
 5   TravelRange            759 non-null    float64
 6   ServiceCeiling         759 non-null    float64
 7   RateOfClimb            759 non-null    float64
 8   MaxTakeOffWeight       759 non-null    float64
 9   MaxPayload             759 non-null    float64
 10  FuelTankCapacity       759 non-null    float64
 11  ExteriorLength         759 non-null    float64
 12  WingSpanRotorDiameter  759 non-null    float64
dtypes: float64(10), int64(1), object(2)
memory usage: 77.2+ KB


Depois de processar e transformar cuidadosamente os dados, criamos com sucesso uma nova tabela que contém 759 linhas e 13 colunas, todas completamente limpas, padronizadas e verificadas para ter os tipos de dados corretos. Para confirmar a precisão de nossos dados, podemos examinar algumas das diferenças entre as tabelas nos exemplos abaixo:  

#### Primeira linha da tabela antiga:

In [10]:
df.head(1)

model                                  Avionics:  \
0  Boeing 787-10 Dreamliner  Rockwell Collins next generation avionics   

                  \n            Engine: \n            \
0  \n            2x Rolls-Royce Trent 1000 or Gen...   

                         Power:          Max Cruise Speed:  \
0  76,000 pound-force            \n              515 knots   

                  Approach Speed (Vref):  \
0  \n              140 knots               

  \n              Travel range:\n              \
0          \n            6,430 Nautical Miles   

                               Fuel Economy:  \
0  \n            0.22 nautical mile / gallon   

                      Service Ceiling:                    Rate of Climb:  ...  \
0  \n            43,000 feet            \n            3000 feet / minute  ...   

              Seats - Business Class:          Seats - First Class:  \
0  \n            290 seats              \n            0               

                                       Cabin Height:  \
0  \n              2.5 metre - 8.20 feet               

                                      Cabin Width:  \
0  \n            5.49 metre - 18.01 feet             

                                    Cabin Length:  \
0  \n            62 metre - 203.41 feet             

                                   Exterior Length:  \
0  \n            68.3 metre - 224.08 feet             

                                       Tail height:  \
0  \n            17.02 metre - 55.84 feet             

                                Fuselage Diameter:  \
0  \n            5.77 metre - 18.93 feet             

                         Wing Span / Rotor Diameter:  \
0  \n            60.12 metre - 197.24 feet             

                               Wing Tips:  
0  \n            Raked Wingtips            

[1 rows x 28 columns]

#### Primeira linha da nova tabela:

In [11]:
new_df.head(1)

Model    Power Unit  EnginesQty  MaxCruiseSpeed  \
0  Boeing 787-10 Dreamliner  76000.0  lbf           2           515.0   

   TravelRange  ServiceCeiling  RateOfClimb  MaxTakeOffWeight  MaxPayload  \
0       6430.0         43000.0       3000.0          254011.0     57277.0   

   FuelTankCapacity  ExteriorLength  WingSpanRotorDiameter  
0           33399.0          224.08                 197.24

In [12]:
#Salvar a nova tabela
#new_df.to_csv("../Files/data.csv", index=False)